### PR TITLE
Lomap atommapper

### LINKED ISSUE #9: Wrap Lomap MCS as LomapAtomMapper

### PR DESCRIPTION
**Pull request by richardjgowers**
fixes #9

### CODE REVIEW

**Comment by richardjgowers** [comment]
this is based off #30

File: openfe/setup/lomap_mapper.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+
+"""
+The MCS class from Lomap shamelessly wrapped and used here to match our API.
+
+"""
+from . import AtomMapper, AtomMapping
+
+import lomap
```
Comment (IAlibay) [inline openfe/setup/lomap_mapper.py:10]: Is this based on your refactored version of lomap?

**Comment by dwhswenson** [comment]
@richardjgowers : If we're going have atom mappers be any callable, maybe the best approach is for you to remove `atommapper.py` and `test_atommapper.py` as part of this PR, and then I just close #30? Rebase or squash merge after?

If I remove them in #30, then there's no PR to merge. (I think that might even auto-close it.)

**Comment by richardjgowers** [comment]
@dwhswenson we have reduced `AtomMapper` to just any callable, but having slept on it, I think the baseclass & inheritance works as quasi-documentation to show devs what needs implementing.

File: openfe/setup/lomap_mapper.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        self.max3d = max3d
+
+    def _mappings_generator(self, mol1, mol2):
+        mcs = lomap.mcs.MCS(mol1, mol2, time=self.time, threed=self.threed,
+                            max3d=self.max3d)
+        # TODO: Once Lomap scorers exist, we'll want to keep a cache of
+        #       these mcs objects ({mapping: mcs}), so we can later query the
+        #       mcs that made a particular mapping to retrieve scores.
+
+        mapping_string = mcs.all_atom_match_list()
```
Comment (IAlibay) [inline openfe/setup/lomap_mapper.py:45]: This is quite ugly - can we upstream this and get a method to return a dictionary in lomap? Would be useful for anyone else using lomap in the future.

**Review by IAlibay — CHANGES_REQUESTED** [review]
some minor comments - needs tests

File: openfe/setup/lomap_mapper.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                 max3d: float = 1000.0):
+        """Wraps the MCS atom mapper from Lomap.
+
+        Kwargs are passed directly to the MCS class from Lomap for each mapping
+        created
+
+        Parameters
+        ----------
+        time : int, optional
+          timeout of MCS algorithm, passed to RDKit
```
[the remark this answers is not in the document]
Comment (richardjgowers) [reply openfe/setup/lomap_mapper.py:25]: I need to actually double check how rdkit uses this, but I assume it's seconds yes

File: openfe/setup/lomap_mapper.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+          default 20
+        threed : bool, optional
```
[the remark this answers is not in the document]
Comment (richardjgowers) [reply openfe/setup/lomap_mapper.py:27]: it's passing on the lomap option, so yes it's weird but it's probably the most viable name

File: openfe/setup/lomap_mapper.py (hunk 2 above)
Comment (richardjgowers) [reply openfe/setup/lomap_mapper.py:45]: possible yes

File: openfe/setup/lomap_mapper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (IAlibay) [reply openfe/setup/lomap_mapper.py:25]: "Use the timeout parameter to stop the search after the given number of seconds (wall-clock seconds, not CPU seconds)"

File: openfe/setup/lomap_mapper.py (hunk 2 above)
Comment (IAlibay) [reply openfe/setup/lomap_mapper.py:45]: issue raised

File: openfe/setup/lomap_mapper.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        """Wraps the MCS atom mapper from Lomap.
+
+        Kwargs are passed directly to the MCS class from Lomap for each mapping
+        created
+
+        Parameters
+        ----------
+        time : int, optional
+          timeout of MCS algorithm, passed to RDKit
+          default 20
```
[the remark this answers is not in the document]
Comment (richardjgowers) [reply openfe/setup/lomap_mapper.py:26]: I'm gonna push back on this, it's kinda annoying to have to jump up and down to see what the default is in the signature, although it's annoying that docs and code could fall out of sync

[the remark this answers is not in the document]
Comment (IAlibay) [reply openfe/setup/lomap_mapper.py:26]: Things falling out of sync is probably the biggest motivator here, but I also know what you mean re: reading order. As long as we're bothered enough to keep things in sync I'm happy. Others might feel differently.

**Comment by richardjgowers** [comment]
Still to go; go back to Lomap original, check outputs then encode them in tests here

File: openfe/tests/setup/test_lomap_atommapper.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    mapper = LomapAtomMapper()
+
+    mapping_gen = mapper.suggest_mappings(mol1, mol2)
+
+    mapping = next(mapping_gen)
+    assert isinstance(mapping, openfe.setup.AtomMapping)
+    # methylcyclohexane to toluene is a 1:1 mapping between all atoms
+    # so 7 values should be present
+    assert len(mapping.mol1_to_mol2) == 7
```
Comment (IAlibay) [inline openfe/tests/setup/test_lomap_atommapper.py:21]: As discussed privately - if you can check the contents themselves that'd be great.

Possibly the only other thing is maybe there should be more cases being checked here, but it's up to you. Something like a "no atoms map" case could be good to test.

Comment (IAlibay) [reply openfe/tests/setup/test_lomap_atommapper.py:21]: If we're waiting on fixing lomap - shall we raise an issue for expanding the tests at a later point?

**Review by IAlibay — APPROVED** [review]
See comments - I'm happy to approve on the basis that we will expand tests in the future.

**Comment by IAlibay** [comment]
> So thinking on this, I'll put the regression tests into the Lomap repo, then have very basic smoke tests here.

Sounds good to me. We only really need to check for edge cases in the lomap returns tbh (e.g. what happens to that string split if there's no matching atoms)

File: openfe/tests/setup/test_lomap_atommapper.py (hunk 6 above)
Comment (richardjgowers) [reply openfe/tests/setup/test_lomap_atommapper.py:22]: lomap regression tests started, so this isn't a here problem

**Comment by richardjgowers** [comment]
@IAlibay added the failed mapping case too.  Should be good to go

File: openfe/tests/setup/test_lomap_atommapper.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    mapping_gen = mapper.suggest_mappings(mol1, mol2)
+
+    _ = next(mapping_gen)
+    with pytest.raises(StopIteration):
+        next(mapping_gen)
+
+
+def test_bad_mapping(lomap_basic_test_files):
+    toluene = lomap_basic_test_files['toluene']
+    NigelTheNitrogen = Chem.MolFromSmiles('N')
```
Comment (IAlibay) [inline openfe/tests/setup/test_lomap_atommapper.py:42]: personified elements are best elements

**Review by IAlibay — APPROVED** [review]
